### PR TITLE
Change fill_NA behaviour in `prepare_data()`

### LINKED ISSUE #82: Fill NAs with 0s by default in prepare_data.incidence2
If a default value of FALSE is more likely to cause issues, why not set it to `TRUE` and throw a message informing the user about the operation? In fact, I would argue that this not be an argument at all but rather operate under a value of `TRUE` assumption, which should be documented as part of the data argument and discussed in the details section of the function.

_Originally posted by @jamesmbaazam in https://github.com/epiverse-trace/cfr/pull/78#discussion_r1320086543_

### PR DESCRIPTION
**Pull request by pratikunterwegs**
This PR makes some small changes to `prepare_data()` to fix #82, and to allow mocked testing of an error message when {incidence2} is not installed.

### CODE REVIEW

File: R/prepare_data.R [hunk 1]
```diff
@@ -1,3 +1,18 @@
...
+#' @description A small internal function to check whether a required package
+#' is installed.
+#' This function is primarily intended to be used in mocking when testing for
+#' appropriate error messages from functions that rely on a package without
+#' that package being imported in {cfr}.
+#' The only current use case is [prepare_data.incidence2()].
+#' @param x A package name as a string.
+#' @return Quietly returns a logical for whether the package passed in `x` is
+#' available locally.
+#' @keywords internal
```
Comment (joshwlambert) [inline R/prepare_data.R:11]: Can optionally also add `@noRd` if you didn't want a file in the manual.

Comment (pratikunterwegs) [reply R/prepare_data.R:11]: Thanks, will do.

File: R/prepare_data.R [hunk 2]
```diff
@@ -104,12 +112,13 @@ prepare_data <- function(data, ...) {
...
-                                    fill_NA = FALSE,
+                                    fill_NA = TRUE,
                                     ...) {
-  # check for {incidence2} and error if not available
-  if (!requireNamespace("incidence2", quietly = TRUE)) {
-    stop("Install package 'incidence2' to prepare <incidence2> class data")
-  }
+  # Check whether incidence2 is installed using internal function
+  stopifnot(
+    "Install package <incidence2> to prepare <incidence2> data" =
```
Comment (joshwlambert) [inline R/prepare_data.R:119]:
```suggestion
    "Install package {incidence2} to prepare <incidence2> data" =
```

File: R/prepare_data.R [hunk 3]
```diff
@@ -1,3 +1,18 @@
+#' @title Check whether a package is installed
+#' @description A small internal function to check whether a required package
+#' is installed.
+#' This function is primarily intended to be used in mocking when testing for
+#' appropriate error messages from functions that rely on a package without
+#' that package being imported in {cfr}.
+#' The only current use case is [prepare_data.incidence2()].
+#' @param x A package name as a string.
```
Comment (joshwlambert) [inline R/prepare_data.R:8]: As this argument is being directly passed to `requireNamespace()` you could use the same argument name (`package`) and then in the documentation use `@inheritParams base::requireNamespace` to save yourself the effort. Minor change either way, and there is nothing wrong with what is there now.

File: R/prepare_data.R [hunk 4]
```diff
@@ -139,6 +148,12 @@ prepare_data.incidence2 <- function(data, cases_variable = "cases",
     data,
     fill = ifelse(fill_NA, 0, NA_integer_)
   )
+  if (fill_NA) {
+    message(
```
Comment (joshwlambert) [inline R/prepare_data.R:152]: I haven't given it a lot of thought, or used the function interactively so could definitely be wrong, but this feels like a `warning()` might be more appropriate, depending on how much the calculation is dependent on missing values being assumed zero.

**Review by joshwlambert — APPROVED** [review]
Nice work! I especially like the use of mock testing.

Changes look good to me. Would also be good to get a quick look from @jamesmbaazam as he is quoted in the issue being resolved. 

Another point, not about the code, it would be good to have an informative PR title, even though it relates back to the issue which contains the details, having it also be self-describing is useful for time saving.

File: R/prepare_data.R (hunk 4 above)
Comment (pratikunterwegs) [reply R/prepare_data.R:152]: No calculations in this function depend on zero filling, but downstream ones expect that there are no NAs. So the only choice is to fill with zeroes or to filter out the data. I'm not sure what to do here, there are arguments for both

**Comment by joshwlambert** [comment]
Taking a closer look at the code around the diffs in this PR and building on what I interpret from issue #82, I'm wondering what role the `fill_NA` argument has in the function, and wonder whether it should be removed. The `prepare_date()` function is titled: "Prepare common epidemiological data formats for CFR estimation", but if `fill_NA = FALSE` then the data is not prepared for a CFR estimation function (as per [L88 of the vignette](https://github.com/epiverse-trace/cfr/blob/main/vignettes/data_from_incidence2.Rmd#L88)). The default behaviour could be to check if `NA`s are present, and if they are found replace with `0` and only with replacement give a message or warning to the user. The only justification for using `prepare_data()` with `fill_NA = FALSE` is if it were part of tidyversesque pipeline that didn't terminate in a CFR calculation, but that seems unlikely.

**Comment by jamesmbaazam** [comment]
I think `prepare_data()` here has a similar goal to [`EpiNow2::create_clean_reported_cases()`](https://github.com/epiforecasts/EpiNow2/blob/af8c1f3fc3c844497a246c60b934ccad5420c362/R/create.R#L22) in that it replaces implicit missing dates with 0 by default and only gives the user some flexibility with the `zero_frequency` argument. It's a similar thing I raise in #82 that `fill_NA` should be removed.

**Comment by pratikunterwegs** [comment]
Thanks for taking a look. The idea with this function is primarily to help with the aggregation and pivoting of incidence2 objects. People fall into two camps with zero filling, and I'm trying to cater to both. Users who are certain that their data are robust to zero filling can stick with the default option. Users who would prefer to be conservative and retain NAs can do that, although they would require intermediate steps (removing NAs and splitting into contiguous date blocks) before passing the data to `cfr_*()`.

File: R/prepare_data.R (hunk 4 above)
Comment (pratikunterwegs) [reply R/prepare_data.R:153]: I've left this as is for now, we can always upgrade this to a warning in future versions.

File: R/prepare_data.R (hunk 3 above)
Comment (pratikunterwegs) [reply R/prepare_data.R]: Implemented in https://github.com/epiverse-trace/cfr/pull/100/commits/e877c4c72e1ab3b06fef7ff5367e6a273f03fca7

File: R/prepare_data.R (hunk 1 above)
Comment (pratikunterwegs) [reply R/prepare_data.R:12]: Added in https://github.com/epiverse-trace/cfr/pull/100/commits/e877c4c72e1ab3b06fef7ff5367e6a273f03fca7

**Comment by pratikunterwegs** [comment]
Thanks both for taking a look, this will be merged once checks pass.

**Pull request merged**